src/purge.rs

```diff
@@ -323,32 +332,56 @@ where
                 m.region_id()
             };
 
-            // FIXME: This code makes my brain hurt.
-            let mut cursor = 0;
-            while cursor < entry_indexes.len() {
-                let entry =
-                    read_entry_bytes_from_file(self.pipe_log.as_ref(), &entry_indexes[cursor])?;
-                total_size += entry.len();
-                entries.push(entry);
-                if total_size > MAX_REWRITE_BATCH_BYTES {
-                    let mut take_entries = Vec::with_capacity(expect_rewrites_per_memtable);
-                    std::mem::swap(&mut take_entries, &mut entries);
-                    let mut take_entry_indexes = entry_indexes.split_off(cursor + 1);
-                    std::mem::swap(&mut take_entry_indexes, &mut entry_indexes);
-                    log_batch.add_raw_entries(region_id, take_entry_indexes, take_entries)?;
-                    self.rewrite_impl(&mut log_batch, rewrite, false)?;
-                    total_size = 0;
-                    cursor = 0;
-                } else {
-                    cursor += 1;
+            // Split the entries into smaller chunks, so that we don't OOM, and the
+            // compression overhead is not too high.
+            let mut chunks = vec![Vec::new()];
+            let mut current_size = 0;
+            debug_assert!(log_batch.approximate_size() <= max_batch_bytes());
+            for ei in entry_indexes {
+                current_size += ei.entry_len as usize;
```

Comment:
What's the difference between `entry.len()` and `ei.entry_len`?

Reply:
They should be the same, the `read_entry_bytes_from_file` makes sure of it.